[GSoC] Implemented New notification preference UI

prateek-singh-3212 opened the pull request:
## Pull Request template

## Purpose / Description
The user is not able to manage the notification preference easily and we have to provide an additional feature to the notification System and make notification system bug free.
GSoC Proposal: https://docs.google.com/document/d/1ezQsKJxXm_-q4d5Kpafb5WHKuy3hApcPo0RXO7BULIc/edit?usp=drivesdk



https://user-images.githubusercontent.com/76490368/195495580-766c72fd-f448-4fd2-9328-5593df67c2f2.mp4


## Fixes
Issue: #4944

## Approach
 I have implemented a new notification preference UI to manage all notification settings related to a single deck in one place

## How Has This Been Tested?

Tested on Samsung M52

**Note: Please ignore the first 2 commits as they are part of PR: #12234**

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

david-allison commented:
This doesn't build.

Could you please re-link the video or screenshots, they're mentioned in the PR header, but I don't see them

prateek-singh-3212 commented:
> This doesn't build.

The build is failing because of Timepicker 'currentHour' method is depricated above API 23. but I was unable to find the correct replacement for that...

david-allison commented:
![Screenshot 2022-10-13 at 20 13 26](https://user-images.githubusercontent.com/62114487/195688948-ade84b5c-a31c-4fa2-8e15-f0179bb7ad35.png)

david-allison commented:
Move the accessor to our `Compat` infrastructure and handle it there

prateek-singh-3212 commented:
@ankidroid/gsoc-mentor I implemented the complete system. You can check this branch [BRANCH](https://github.com/prateek-singh-3212/Anki-Android/tree/Notification_Final_Testing)

I am attaching the video of the same. The timezone change is also working (Please let me know If you want me to add a timezone change video too) 

https://user-images.githubusercontent.com/76490368/197379174-d7f52c61-53b3-4aa1-8202-0dd7ba063d35.mp4


**NOTE: use this tag `package:com.ichi2.anki notification` to view logs**

1. Logs when I added the `@r$     Biology          ANIMAL KINGDOM` for notification
```
2022-10-23 12:38:27.627 21659-21892 Notificati...tification com.ichi2.anki                       D  CreateScheduledDeckNotification-> did: 1659703391318 hour: 12 min: 40
2022-10-23 12:38:27.629 21659-21892 NotificationHelper      com.ichi2.anki                       D  Scheduled time is not yet passed.
2022-10-23 12:38:27.629 21659-21892 NotificationHelper      com.ichi2.anki                       D  Creating new time deck data 
2022-10-23 12:38:27.726 21659-22366 Notificati...tification com.ichi2.anki                       D  Next trigger time 1666509027628
2022-10-23 12:38:27.726 21659-22366 NotificationHelper      com.ichi2.anki                       D  Starting work manager with initial delay 119902 force: true
```
2. Logs when I added the `Biology` for notification.
```
2022-10-23 12:38:40.570 21659-22366 Notificati...tification com.ichi2.anki                       D  CreateScheduledDeckNotification-> did: 1665512292295 hour: 12 min: 39
2022-10-23 12:38:40.586 21659-22366 NotificationHelper      com.ichi2.anki                       D  Scheduled time is not yet passed.
2022-10-23 12:38:40.586 21659-22366 NotificationHelper      com.ichi2.anki                       D  Scheduled time will come today only. And it will come before the current scheduled deck.
2022-10-23 12:38:40.591 21659-21892 Notificati...tification com.ichi2.anki                       D  Next trigger time 1666508980586
2022-10-23 12:38:40.591 21659-21892 NotificationHelper      com.ichi2.anki                       D  Starting work manager with initial delay 59995 force: true
```

3. Logs when System is triggering the notification for  `Biology`
```
2022-10-23 12:39:40.605 21659-22366 NotificationWorker      com.ichi2.anki                       D  NotificationManagerWorker: Worker status -> STARTED
2022-10-23 12:39:40.686 21659-22366 NotificationWorker      com.ichi2.anki                       D  Processing single deck notification...
2022-10-23 12:39:40.692 21659-22366 NotificationWorker      com.ichi2.anki                       D  top decks: [@r$     Biology          ANIMAL KINGDOM, 1659703391318, 0, 0, 20, [], Biology, 1665512292295, 0, 0, 20, [Biology::01: Characteristics and classification, 1665512292296, 0, 0, 18, [], Biology::02: Organisation of the organism, 1665512292299, 0, 0, 14, [], Biology::03: Movement in and out of cells, 1665512292302, 0, 0, 14, [], Biology::04: Biological molecules, 1665512292306, 0, 0, 19, [], Biology::05: Enzymes, 1665512292310, 0, 0, 9, [], Biology::06: Plant nutrition, 1665512292312, 0, 0, 15, [], Biology::07: Human nutrition, 1665512292316, 0, 0, 20, [], Biology::08: Transport in plants, 1665512292326, 0, 0, 16, [], Biology::09: Transport in animals, 1665512292331, 0, 0, 20, [], Biology::10: Diseases and immunity, 1665512292342, 0, 0, 18, [], Biology::11: Gas exchange in humans, 1665512292346, 0, 0, 20, [], Biology::12: Respiration, 1665512292352, 0, 0, 12, [], Biology::13: Excretion in humans, 1665512292355, 0, 0, 20, [], Biology::14: Coordination and response, 1665512292361, 0, 0, 20, [], Biology::15: Drugs, 1665512292375, 0, 0, 18, [], Biology::16: Reproduction, 1665512292379, 0, 0, 20, [], Biology::17: Inheritance, 1665512292392, 0, 0, 20, [], Biology::18: Variation and selection, 1665512292397, 0, 0, 19, [], Biology::19: Organisms and their environment, 1665512292401, 0, 0, 20, [], Biology::20: Biotechnology and genetic engineering, 1665512292409, 0, 0, 9, [], Biology::21: Human influences on ecosystems, 1665512292411, 0, 0, 20, []], Biology::01: Characteristics and classification, 1665512292296, 0, 0, 18, [], Biology::02: Organisation of the organism, 1665512292299, 0, 0, 14, [], Biology::03: Movement in and out of cells, 1665512292302, 0, 0, 14, [], Biology::04: Biological molecules, 1665512292306, 0, 0, 19, [], Biology::05: Enzymes, 1665512292310, 0, 0, 9, [], Biology::06: Plant nutrition, 1665512292312, 0, 0, 15, [], Biology::07: Human nutrition, 1665512292316, 0, 0, 20, [], Biology::08: Transport in plants, 1665512292326, 0, 0, 16, [], Biology::09: Transport in animals, 1665512292331, 0, 0, 20, [], Biology::10: Diseases and immunity, 1665512292342, 0, 0, 18, [], Biology::11: Gas exchange in humans, 1665512292346, 0, 0, 20, [], Biology::12: Respiration, 1665512292352, 0, 0, 12, [], Biology::13: Excretion in humans, 1665512292355, 0, 0, 20, [], Biology::14: Coordination and response, 1665512292361, 0, 0, 20, [], Biology::15: Drugs, 1665512292375, 0, 0, 18, [], Biology::16: Reproduction, 1665512292379, 0, 0, 20, [], Biology::17: Inheritance, 1665512292392, 0, 0, 20, [], Biology::18: Variation and selection, 1665512292397, 0, 0, 19, [], Biology::19: Organisms and their environment, 1665512292401, 0, 0, 20, [], Biology::20: Biotechnology and genetic engineering, 1665512292409, 0, 0, 9, [], Biology::21: Human influences on ecosystems, 1665512292411, 0, 0, 20, [], Default, 1, 0, 0, 0, [], Kieler Basisdeck Medizin, 1664979398946, 0, 3, 20, [], Math, 1664979441850, 4, 1, 20, [Math::Derivative, 1664979441866, 0, 0, 18, [], Math::Integration, 1664979441851, 4, 0, 20, [Math::Integration::StandartIntegration, 1664979441852, 0, 0, 15, [], Math::Integration::Trig-1, 1664979441858, 4, 0, 0, [], Math::Integration::Trig-2, 1664979441862, 0, 0, 10, []], Math::Trigonometry, 1664979441873, 0, 1, 20, [Math::Trigonometry::AngleTransf., 1664979441874, 0, 1, 7, [], Math::Trigonometry::General, 1664979441878, 0, 0, 8, [], Math::Trigonometry::HalfDoubleAngles, 1664979441881, 0, 0, 8, [], Math::Trigonometry::Product, 1664979441885, 0, 0, 6, [], Math::Trigonometry::Sum-1, 1664979441888, 0, 0, 8, [], Math::Trigonometry::Sum-2, 1664979441891, 0, 0, 8, []]], Math::Derivative, 1664979441866, 0, 0, 18, [], Math::Integration, 1664979441851, 4, 0, 20, [Math::Integration::StandartIntegration, 1664979441852, 0, 0, 15, [], Math::Integration::Trig-1, 1664979441858, 4, 0, 0, [], Math::Integration::Trig-2, 1664979441862, 0, 0, 10, []], Math::Integration::StandartIntegration, 1664979441852, 0, 0, 15, [], Math::Integration::Trig-1, 1664979441858, 4, 
2022-10-23 12:39:40.693 21659-22366 NotificationWorker      com.ichi2.anki                       D  0, 0, [], Math::Integration::Trig-2, 1664979441862, 0, 0, 10, [], Math::Trigonometry, 1664979441873, 0, 1, 20, [Math::Trigonometry::AngleTransf., 1664979441874, 0, 1, 7, [], Math::Trigonometry::General, 1664979441878, 0, 0, 8, [], Math::Trigonometry::HalfDoubleAngles, 1664979441881, 0, 0, 8, [], Math::Trigonometry::Product, 1664979441885, 0, 0, 6, [], Math::Trigonometry::Sum-1, 1664979441888, 0, 0, 8, [], Math::Trigonometry::Sum-2, 1664979441891, 0, 0, 8, []], Math::Trigonometry::AngleTransf., 1664979441874, 0, 1, 7, [], Math::Trigonometry::General, 1664979441878, 0, 0, 8, [], Math::Trigonometry::HalfDoubleAngles, 1664979441881, 0, 0, 8, [], Math::Trigonometry::Product, 1664979441885, 0, 0, 6, [], Math::Trigonometry::Sum-1, 1664979441888, 0, 0, 8, [], Math::Trigonometry::Sum-2, 1664979441891, 0, 0, 8, []] 
2022-10-23 12:39:40.693 21659-22366 NotificationWorker      com.ichi2.anki                       D   deckIds: [1665512292295] 
2022-10-23 12:39:40.693 21659-22366 NotificationWorker      com.ichi2.anki                       D   deckNotification: [Biology, 1665512292295, 0, 0, 20, [Biology::01: Characteristics and classification, 1665512292296, 0, 0, 18, [], Biology::02: Organisation of the organism, 1665512292299, 0, 0, 14, [], Biology::03: Movement in and out of cells, 1665512292302, 0, 0, 14, [], Biology::04: Biological molecules, 1665512292306, 0, 0, 19, [], Biology::05: Enzymes, 1665512292310, 0, 0, 9, [], Biology::06: Plant nutrition, 1665512292312, 0, 0, 15, [], Biology::07: Human nutrition, 1665512292316, 0, 0, 20, [], Biology::08: Transport in plants, 1665512292326, 0, 0, 16, [], Biology::09: Transport in animals, 1665512292331, 0, 0, 20, [], Biology::10: Diseases and immunity, 1665512292342, 0, 0, 18, [], Biology::11: Gas exchange in humans, 1665512292346, 0, 0, 20, [], Biology::12: Respiration, 1665512292352, 0, 0, 12, [], Biology::13: Excretion in humans, 1665512292355, 0, 0, 20, [], Biology::14: Coordination and response, 1665512292361, 0, 0, 20, [], Biology::15: Drugs, 1665512292375, 0, 0, 18, [], Biology::16: Reproduction, 1665512292379, 0, 0, 20, [], Biology::17: Inheritance, 1665512292392, 0, 0, 20, [], Biology::18: Variation and selection, 1665512292397, 0, 0, 19, [], Biology::19: Organisms and their environment, 1665512292401, 0, 0, 20, [], Biology::20: Biotechnology and genetic engineering, 1665512292409, 0, 0, 9, [], Biology::21: Human influences on ecosystems, 1665512292411, 0, 0, 20, []]]
2022-10-23 12:39:40.693 21659-22366 NotificationWorker      com.ichi2.anki                       D  Firing deck notification for did -> 1665512292295
2022-10-23 12:39:40.707 21659-22366 NotificationWorker      com.ichi2.anki                       D  Firing all deck notification.
2022-10-23 12:39:40.710 21659-22366 NotificationWorker      com.ichi2.anki                       D  NotificationManagerWorker: Worker status -> FINISHED
2022-10-23 12:39:40.710 21659-22366 NotificationWorker      com.ichi2.anki                       D  Task Completed. Rescheduling...
2022-10-23 12:39:40.711 21659-22366 NotificationWorker      com.ichi2.anki                       D  Next trigger time 1666509027628 Initial Difference 46917
2022-10-23 12:39:40.711 21659-22366 NotificationHelper      com.ichi2.anki                       D  Starting work manager with initial delay 46917 force: true
```

4 Logs when System is triggering the notification for  `@r$     Biology          ANIMAL KINGDOM`
```
2022-10-23 12:40:27.654 21659-22366 NotificationWorker      com.ichi2.anki                       D  NotificationManagerWorker: Worker status -> STARTED
2022-10-23 12:40:27.726 21659-22366 NotificationWorker      com.ichi2.anki                       D  Processing single deck notification...
2022-10-23 12:40:27.731 21659-22366 NotificationWorker      com.ichi2.anki                       D  top decks: [@r$     Biology          ANIMAL KINGDOM, 1659703391318, 0, 0, 20, [], Biology, 1665512292295, 0, 0, 20, [Biology::01: Characteristics and classification, 1665512292296, 0, 0, 18, [], Biology::02: Organisation of the organism, 1665512292299, 0, 0, 14, [], Biology::03: Movement in and out of cells, 1665512292302, 0, 0, 14, [], Biology::04: Biological molecules, 1665512292306, 0, 0, 19, [], Biology::05: Enzymes, 1665512292310, 0, 0, 9, [], Biology::06: Plant nutrition, 1665512292312, 0, 0, 15, [], Biology::07: Human nutrition, 1665512292316, 0, 0, 20, [], Biology::08: Transport in plants, 1665512292326, 0, 0, 16, [], Biology::09: Transport in animals, 1665512292331, 0, 0, 20, [], Biology::10: Diseases and immunity, 1665512292342, 0, 0, 18, [], Biology::11: Gas exchange in humans, 1665512292346, 0, 0, 20, [], Biology::12: Respiration, 1665512292352, 0, 0, 12, [], Biology::13: Excretion in humans, 1665512292355, 0, 0, 20, [], Biology::14: Coordination and response, 1665512292361, 0, 0, 20, [], Biology::15: Drugs, 1665512292375, 0, 0, 18, [], Biology::16: Reproduction, 1665512292379, 0, 0, 20, [], Biology::17: Inheritance, 1665512292392, 0, 0, 20, [], Biology::18: Variation and selection, 1665512292397, 0, 0, 19, [], Biology::19: Organisms and their environment, 1665512292401, 0, 0, 20, [], Biology::20: Biotechnology and genetic engineering, 1665512292409, 0, 0, 9, [], Biology::21: Human influences on ecosystems, 1665512292411, 0, 0, 20, []], Biology::01: Characteristics and classification, 1665512292296, 0, 0, 18, [], Biology::02: Organisation of the organism, 1665512292299, 0, 0, 14, [], Biology::03: Movement in and out of cells, 1665512292302, 0, 0, 14, [], Biology::04: Biological molecules, 1665512292306, 0, 0, 19, [], Biology::05: Enzymes, 1665512292310, 0, 0, 9, [], Biology::06: Plant nutrition, 1665512292312, 0, 0, 15, [], Biology::07: Human nutrition, 1665512292316, 0, 0, 20, [], Biology::08: Transport in plants, 1665512292326, 0, 0, 16, [], Biology::09: Transport in animals, 1665512292331, 0, 0, 20, [], Biology::10: Diseases and immunity, 1665512292342, 0, 0, 18, [], Biology::11: Gas exchange in humans, 1665512292346, 0, 0, 20, [], Biology::12: Respiration, 1665512292352, 0, 0, 12, [], Biology::13: Excretion in humans, 1665512292355, 0, 0, 20, [], Biology::14: Coordination and response, 1665512292361, 0, 0, 20, [], Biology::15: Drugs, 1665512292375, 0, 0, 18, [], Biology::16: Reproduction, 1665512292379, 0, 0, 20, [], Biology::17: Inheritance, 1665512292392, 0, 0, 20, [], Biology::18: Variation and selection, 1665512292397, 0, 0, 19, [], Biology::19: Organisms and their environment, 1665512292401, 0, 0, 20, [], Biology::20: Biotechnology and genetic engineering, 1665512292409, 0, 0, 9, [], Biology::21: Human influences on ecosystems, 1665512292411, 0, 0, 20, [], Default, 1, 0, 0, 0, [], Kieler Basisdeck Medizin, 1664979398946, 0, 3, 20, [], Math, 1664979441850, 4, 1, 20, [Math::Derivative, 1664979441866, 0, 0, 18, [], Math::Integration, 1664979441851, 4, 0, 20, [Math::Integration::StandartIntegration, 1664979441852, 0, 0, 15, [], Math::Integration::Trig-1, 1664979441858, 4, 0, 0, [], Math::Integration::Trig-2, 1664979441862, 0, 0, 10, []], Math::Trigonometry, 1664979441873, 0, 1, 20, [Math::Trigonometry::AngleTransf., 1664979441874, 0, 1, 7, [], Math::Trigonometry::General, 1664979441878, 0, 0, 8, [], Math::Trigonometry::HalfDoubleAngles, 1664979441881, 0, 0, 8, [], Math::Trigonometry::Product, 1664979441885, 0, 0, 6, [], Math::Trigonometry::Sum-1, 1664979441888, 0, 0, 8, [], Math::Trigonometry::Sum-2, 1664979441891, 0, 0, 8, []]], Math::Derivative, 1664979441866, 0, 0, 18, [], Math::Integration, 1664979441851, 4, 0, 20, [Math::Integration::StandartIntegration, 1664979441852, 0, 0, 15, [], Math::Integration::Trig-1, 1664979441858, 4, 0, 0, [], Math::Integration::Trig-2, 1664979441862, 0, 0, 10, []], Math::Integration::StandartIntegration, 1664979441852, 0, 0, 15, [], Math::Integration::Trig-1, 1664979441858, 4, 
2022-10-23 12:40:27.731 21659-22366 NotificationWorker      com.ichi2.anki                       D  0, 0, [], Math::Integration::Trig-2, 1664979441862, 0, 0, 10, [], Math::Trigonometry, 1664979441873, 0, 1, 20, [Math::Trigonometry::AngleTransf., 1664979441874, 0, 1, 7, [], Math::Trigonometry::General, 1664979441878, 0, 0, 8, [], Math::Trigonometry::HalfDoubleAngles, 1664979441881, 0, 0, 8, [], Math::Trigonometry::Product, 1664979441885, 0, 0, 6, [], Math::Trigonometry::Sum-1, 1664979441888, 0, 0, 8, [], Math::Trigonometry::Sum-2, 1664979441891, 0, 0, 8, []], Math::Trigonometry::AngleTransf., 1664979441874, 0, 1, 7, [], Math::Trigonometry::General, 1664979441878, 0, 0, 8, [], Math::Trigonometry::HalfDoubleAngles, 1664979441881, 0, 0, 8, [], Math::Trigonometry::Product, 1664979441885, 0, 0, 6, [], Math::Trigonometry::Sum-1, 1664979441888, 0, 0, 8, [], Math::Trigonometry::Sum-2, 1664979441891, 0, 0, 8, []] 
2022-10-23 12:40:27.731 21659-22366 NotificationWorker      com.ichi2.anki                       D   deckIds: [1659703391318] 
2022-10-23 12:40:27.731 21659-22366 NotificationWorker      com.ichi2.anki                       D   deckNotification: [@r$     Biology          ANIMAL KINGDOM, 1659703391318, 0, 0, 20, []]
2022-10-23 12:40:27.731 21659-22366 NotificationWorker      com.ichi2.anki                       D  Firing deck notification for did -> 1659703391318
2022-10-23 12:40:27.744 21659-22367 NotificationWorker      com.ichi2.anki                       D  Firing all deck notification.
2022-10-23 12:40:27.756 21659-22367 NotificationWorker      com.ichi2.anki                       D  NotificationManagerWorker: Worker status -> FINISHED
2022-10-23 12:40:27.756 21659-22367 NotificationWorker      com.ichi2.anki                       D  Task Completed. Rescheduling...
2022-10-23 12:40:27.758 21659-22367 NotificationWorker      com.ichi2.anki                       D  Next trigger time 1666512627758 Initial Difference 3600000
2022-10-23 12:40:27.758 21659-22367 NotificationHelper      com.ichi2.anki                       D  Starting work manager with initial delay 3600000 force: true
```

This is the system which is built so far. Please let me know the additional thinks which should be added in the **basic notification system** i.e things required in the first relase.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/model/DeckNotification.kt:
```diff
@@ -0,0 +1,57 @@
+import com.ichi2.libanki.Decks
+
+/**
+ * Notification details of particular deck.
```
This comment isn't very helpful

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/model/DeckNotification.kt:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Notification details of particular deck.
+ * */
+data class DeckNotification(
```
I'd prefer the name to reflect that this is about reminding to study. It's hard to give this an exact name but maybe something like `RemindToStudy`. Also I'd ditch `enabled` and rely on presence or absence of instances of this class.

Also, a matter of personal preference, I'd say `deckId`. In my opinion, `did` is a mistake not worth following.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/model/DeckNotification.kt:
```diff
@@ -0,0 +1,57 @@
+) {
+
+    // Empty constructor. Required for jackson to serialize.
+    constructor() : this(
```
I think this is way too simple to serialize.

Serialization is fine for possibly complex objects such as encryption keys, or for objects that aren't supposed to be long-lived. Serialization is nice until you need to change how you store your data, or until some weird thing changes and you can't deserialize. For preferences, especially in such simple cases like this, I'd strongly prefer something more manual. E.g. just encode all of enabled decks and their times in a string.

Also, why jackson istead of kotlinx.serialization?

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/model/DeckNotification.kt:
```diff
@@ -0,0 +1,57 @@
+ * @param started Day on which deck started.
+ * @param completed Deck is done for the day or not.
+ * */
+data class UserNotificationPreference(
```
I have no idea what this entire class is about.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationDatastore.kt:
```diff
@@ -0,0 +1,313 @@
+/*
+* We use TimeOfNotification as key of json object. So it must be string.
+*/
+typealias TimeOfNotification = String
```
Why does this need a new type? Epoch milliseconds should be simply a `Long`, or `EpochMillis` if you wish.

Also, surely jackson can deal with non-string keys? kotlinx does it via storing maps as lists.

Also, when I hear “time of notification” I think of stuff like "HH:MM". This is clearly not it.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationDatastore.kt:
```diff
@@ -0,0 +1,313 @@
+typealias SimultaneouslyTriggeredDeckIds = HashSet<DeckId>
+
+/**
+ * Indicates all notifications that we must eventually trigger. If a notification time is in the past, we must do it as soon as possible, otherwise we must trigger the notification later.
```
This line is way too long. I suggest keeping at around 100 characters max, which is AS default. (My personal favorite is [88](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length).)

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationDatastore.kt:
```diff
@@ -0,0 +1,313 @@
+ * Indicates all notifications that we must eventually trigger. If a notification time is in the past, we must do it as soon as possible, otherwise we must trigger the notification later.
+ * Implemented as map from time to set of deck ids triggered at this time
+ */
+typealias NotificationTodo = TreeMap<TimeOfNotification, SimultaneouslyTriggeredDeckIds>
```
Why is this a new type? Also, why is this a *hash* set specifically?

IMO, you only ever use type aliases:

* to disambiguate things that have the same name,
* to create aliases for long, hard to read types,
* to give a name to an object that could be another type or an interface.

If you think you have the last case, perhaps think of it like this. 

* does it make sense to use another class in this case?
* if yes, can you get away with using a `value class`?
* if yes, but it would be too difficult to implement, use a type alias.

I don't think this case applies here.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationDatastore.kt:
```diff
@@ -0,0 +1,313 @@
+ * Notification which is going to trigger next.
+ * */
+fun NotificationTodo.earliestNotifications(): MutableMap.MutableEntry<TimeOfNotification, SimultaneouslyTriggeredDeckIds>? =
+    firstEntry()
```
Okay, you just created an extension method for all `TreeMap<String, HashSet<Long>>`. This needs to be a class or something.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationDatastore.kt:
```diff
@@ -0,0 +1,313 @@
+ * starting with a 1 that will be a number greater than previous numbers but alphabetically shorter
+ * */
+fun TimeOfNotification.compare(timeOfNotification: TimeOfNotification) =
+    this.toLong().compareTo(timeOfNotification.toLong())
```
...or just use `Long` in the first place

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationDatastore.kt:
```diff
@@ -0,0 +1,313 @@
+ * */
+fun NotificationTodoObject() = NotificationTodo { timeOfNotification1, timeOfNotification2 ->
+    timeOfNotification1.compare(timeOfNotification2)
+}
```
Right, this function here is clear indicator that

Arthur-Milchior replied:
I do not know if you're being ironic here, or if you just didn't finish your sentence.

Arthur-Milchior replied:
I do not know if you're being ironic here, or if you just didn't finish your sentence.

oakkitten replied:
I *had* to write something but I just couldn't bring myself to finish the sentence. But hey you finished yours twice so it evens out!

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationDatastore.kt:
```diff
@@ -0,0 +1,313 @@
+ * Stores the scheduled notification details
+ * This is a singleton class, use [getInstance]
+ * */
+class NotificationDatastore private constructor(val context: Context) {
```
I think this is **extremely** overengineered. I would like this entire class removed. Instead, I would just take all decks and their times and put it in a string such as "123:12:00 456:18:00" and store it in shared preferences under a single key and call it a day. A couple of lines to decode, a couple of lines to encode.

Also, calling async functions like that creates a race condition. If you call a methods such as `putIntSync`, reading the value immediately after may yield the old value.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationDatastore.kt:
```diff
@@ -0,0 +1,313 @@
+     * Stores the Map of time and list of deck ids to Datastore
+     * It stores the data asynchronously.
+     * */
+    suspend fun setTimeDeckData(data: Map<String, HashSet<Long>>) {
```
Did you mean `NotificationTodo` here? Also, why is this saved in the first place?

What is “time deck data”?

prateek-singh-3212 replied:
> Did you mean NotificationTodo here? Also, why is this saved in the first place?

Yes, Actually Their is the error when I pass the notificationTodo to Json Object so I wrote this

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationDatastore.kt:
```diff
@@ -0,0 +1,313 @@
+    * We actually are not blocking the thread. This method throws an exception. It will not create problem for us.
+    * */
+    @Suppress("UNCHECKED_CAST", "BlockingMethodInNonBlockingContext")
+    suspend fun getTimeDeckData(): NotificationTodo? {
```
Method “gets” “time deck data” but returns `NotificationTodo?`.

Method yields a `TreeMap` without a comparator used by the `NotificationTodoObject` thing. Not that `NotificationTodoObject` is used by anything else, but still.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationDatastore.kt:
```diff
@@ -0,0 +1,313 @@
+     * Stores the details of the [notification] scheduling of deck [did]
+     * @return operation successful of not.
+     * */
+    suspend fun setDeckSchedData(did: DeckId, notification: DeckNotification): Boolean {
```
I don't think this function saves deck scheduling data...

prateek-singh-3212 replied:
Why?

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationDatastore.kt:
```diff
@@ -0,0 +1,313 @@
+                    it[datastoreKey],
+                    DeckNotification::class.java
+                )
+            } catch (ex: Exception) {
```
Code two methods above catches `JacksonException` and doesn't send a report. Why is this different?

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationDatastore.kt:
```diff
@@ -0,0 +1,313 @@
+         * Thread safe. We're using this pattern because we need a Context to get an instance.
+         * @return The singleton NotificationDatastore
+         */
+        fun getInstance(context: Context) = instanceInitializedOr(context) {
```
Appreciate the double-checked locking but why does this need to be thread safe?

Also, IIRC companion objects are created in a thread-safe manner so you should just be able to say `INSTANCE = ...`.

Arthur-Milchior replied:
I can't easily find a source for the thread-safe part. At least not on https://kotlinlang.org/docs/object-declarations.html#semantic-difference-between-object-expressions-and-declarations which is the place I'd expect to see such an information

oakkitten replied:
It's a SO answer and not official docs but it makes sense I guess? https://stackoverflow.com/a/30190567

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/NotificationPreferenceAdapter.kt:
```diff
@@ -0,0 +1,234 @@
+import net.ankiweb.rsdroid.BackendFactory
+import timber.log.Timber
+
+class NotificationPreferenceAdapter(val layoutInflater: LayoutInflater, val context: Context) :
```
While I have some comments here, if this works, I suggest keeping it as, or simplifying to bare minimum, so that it can be possibly replaced with the adapter mentioned in #12554

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/NotificationPreferenceAdapter.kt:
```diff
@@ -0,0 +1,234 @@
+    RecyclerView.Adapter<NotificationPreferenceAdapter.ViewHolder>() {
+
+    // Recyclerview Data
+    private val mDeckList: MutableList<TreeNode<AbstractDeckTreeNode>>
```
Just a note; when making adapters, go for `var items: List` and not `val items: MutableList`. Might require more code, but the logic becomes simpler, and it's easier to be make sure that no data is mutated while being read by the view. And you can do diffs.

Also, I thought we don't do the `m` prefixes anymore?

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/NotificationPreferenceAdapter.kt:
```diff
@@ -0,0 +1,234 @@
+        val treeNode = mDeckList[position]
+
+        // Get Notification Time
+        CoroutineScope(Dispatchers.Main).launch {
```
I understand what's going on here, but this is ***very*** problematic. By the time `holder.notificationTime.text` runs, this holder might be referring to a whole new item, so you can have wrong times displayed in wrong positions. A proper way to fix this would be to cancel a job when the holder becomes unbound; a simpler solution would be to check if the holder is still valid. A more proper solution would be to not go async.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/NotificationPreferenceAdapter.kt:
```diff
@@ -0,0 +1,234 @@
+            val deckNotification =
+                NotificationDatastore.getInstance(context).getDeckSchedData(treeNode.value.did)
+            holder.notificationTime.text =
+                deckNotification?.let { "${it.schedHour}:${it.schedMinutes}" } ?: "HH:MM"
```
Time format should not be hardcoded

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/NotificationPreferenceAdapter.kt:
```diff
@@ -0,0 +1,234 @@
+                deckNotification?.let { "${it.schedHour}:${it.schedMinutes}" } ?: "HH:MM"
+        }
+        val rightPadding =
+            holder.deckLayout.resources.getDimension(R.dimen.deck_picker_right_padding).toInt()
```
Cache value in a field for easier reading ~~and speed~~.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/NotificationPreferenceAdapter.kt:
```diff
@@ -0,0 +1,234 @@
+
+        if (treeNode.hasChildren()) {
+            holder.deckExpander.tag = treeNode.value.did
+            holder.deckExpander.setOnClickListener(mDeckExpanderClickListener)
```
Don't do this in `onBindViewHolder`. Set `ViewHolder`-specific listeners in `onCreateViewHolder` or `ViewHolder`'s constructor. You can then get item id/position/item from a view holder, and you can call your stuff from there.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/NotificationPreferenceAdapter.kt:
```diff
@@ -0,0 +1,234 @@
+        mHasSubdecks = false
+        currentDeckId = mCol.decks.current().optLong("id")
+        processNodes(nodes)
+        notifyDataSetChanged()
```
I think this + not having stable IDs destroys all animation?

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/NotificationPreferenceAdapter.kt:
```diff
@@ -0,0 +1,234 @@
+        notifyDataSetChanged()
+    }
+
+    private fun processNodes(nodes: List<TreeNode<AbstractDeckTreeNode>>) {
```
This method is hard to read, I suggest refactoring

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/NotificationPreferenceAdapter.kt:
```diff
@@ -0,0 +1,234 @@
+            if (BackendFactory.defaultLegacySchema) {
+                // If the default deck is empty, hide it by not adding it to the deck list.
+                // We don't hide it if it's the only deck or if it has sub-decks.
+                if (node.value.did == 1L && nodes.size > 1 && !node.hasChildren()) {
```
Maybe use `shouldShowDefaultDeck()`?

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/NotificationPreferenceAdapter.kt:
```diff
@@ -0,0 +1,234 @@
+        indent.minimumWidth = width
+    }
+
+    fun setTimeClickListener(listener: OnClickListener?) {
```
I'd simply allow setting listeners directly, e.g. `adapter.timeClickListener = ...`

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/NotificationPreferenceAdapter.kt:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    init {
+        mDeckList = ArrayList()
```
Join declaration and assignment for `mDeckList`

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/widgets/NotificationPreferenceAdapter.kt:
```diff
@@ -0,0 +1,234 @@
+            R.attr.collapseRef
+        )
+
+        val obtainedAttrs = context.obtainStyledAttributes(attrs)
```
You forgot to recycle the array. Instead, use `usingStyledAttributes` (our method).

oakkitten commented on AnkiDroid/src/main/res/layout/deck_notification_card.xml:
```diff
@@ -0,0 +1,93 @@
+    android:orientation="horizontal"
+    android:paddingEnd="@dimen/deck_picker_right_padding"
+    android:paddingStart="@dimen/deck_picker_left_padding"
+    tools:background="@android:color/holo_orange_light">
```
Orange?...

oakkitten commented on AnkiDroid/src/main/res/layout/deck_notification_card.xml:
```diff
@@ -0,0 +1,93 @@
+            tools:text="Deck name" />
+    </LinearLayout>
+
+    <LinearLayout
```
Why need for linear layout? Add a comment with explanation

oakkitten commented on AnkiDroid/src/main/res/layout/deck_notification_card.xml:
```diff
@@ -0,0 +1,93 @@
+        android:layout_centerVertical="true"
+        android:background="?attr/selectableItemBackground"
+        android:clickable="true"
+        android:contentDescription="@string/deck_picker_counts"
```
I suppose this got here by mistake, also below

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/NotificationsSettingsFragment.kt:
```diff
@@ -15,64 +15,138 @@
+    private lateinit var mDeckListAdapter: NotificationPreferenceAdapter
+    private lateinit var mRecyclerView: RecyclerView
+    private lateinit var mProgressBar: ProgressBar
+    var dueTree: List<TreeNode<AbstractDeckTreeNode>>? = null
```
Why not `lateinit`?

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/NotificationsSettingsFragment.kt:
```diff
@@ -15,64 +15,138 @@
+        savedInstanceState: Bundle?
+    ): View? {
+        val view = inflater.inflate(R.layout.fragment_preference_notification, container, false)
+        initViews(view)
```
I'd suggest inlining this call

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/NotificationsSettingsFragment.kt:
```diff
@@ -15,64 +15,138 @@
+        }
+        mRecyclerView.apply {
+            adapter = mDeckListAdapter
+            layoutManager = LinearLayoutManager(context)
```
I'd just set layout manager in XML

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/NotificationsSettingsFragment.kt:
```diff
@@ -15,64 +15,138 @@
+
+        // Fetch Deck Data
+        TaskManager.launchCollectionTask(
+            CollectionTask.LoadDeckCounts(),
```
As I understand this only needs deck list. Is there a way to not also load counts?

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/NotificationsSettingsFragment.kt:
```diff
@@ -15,64 +15,138 @@
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    fun toggleDeckExpand(did: DeckId) {
```
I do not think that expanding/collapsing decks in this screen should affect the database. Also, isn't “updating DB” doing IO? (Also, is retrieving collapsed status doing IO?)

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/NotificationsSettingsFragment.kt:
```diff
@@ -15,64 +15,138 @@
+            // update stored state
+            val deck: List<TreeNode<DeckDueTreeNode>> = dueTree!! as List<TreeNode<DeckDueTreeNode>>
+            Timber.d(dueTree!!.toString() + " " + did)
+            findInDeckTree(deck, did)?.run {
```
`apply` is more idiomatic I think

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationSheetDialog.kt:
```diff
@@ -0,0 +1,115 @@
+        fun onDialogueDismiss(position: Int)
+    }
+
+    // Delete Deck Notification
```
Hello stray comment, are you lost?

oakkitten commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -436,4 +436,8 @@
     <string name="browser_options_dialog_heading">Browser Options</string>
+
+    <string name="notification_sheet_subdeck">Include Sub Deck</string>
+    <string name="notification_sheet_sticky">Sticky Notification</string>
```
As a user, I don't know what's a sticky notification. As a dev, I'm not sure if this complexity is worth it.

oakkitten reviewed (changes requested):
I haven't had a chance to run the code yet, so I'm not commenting on the behavior or design here. The code has some major issues, and I think parts of this PR are overengineered (too complicated when a simpler solution is possible). I suggest not serializing, and also removing most of async stuff or at least limiting it a few isolated spaced.

prateek-singh-3212 commented:
Thanks, @oakkitten for reviewing this PR. I will do these UI changes shortly. Actually, the feature that we want to provide to the user is not yet decided completely. I will do all the implementation changes that you requested and once the feature is decided then I will write this as production ready.
I created this PR now because I want maintainers to review my approach on which UI is built. Thanks for suggesting it I will do the changes. And feel free to tell me the other approaches (if there is any)

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/NotificationSheetDialog.kt:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    // Save Deck Notification
+    private fun saveDeckNotification(notification: DeckNotification) {
```
I meant absolutely no derision. I see how, if reading it **in isolation**, there could be a chance that it could be taken as such, but do you **really** think that that I wanted to ridicule anyone given the overall tone of my comments?...

BrayanDSO commented:
Codebase changed a lot since this PR was opened, and looks abandoned. It's best to create another one if work is restarted
